cou/apps/auxiliary.py

```diff
@@ -204,10 +205,74 @@ class MysqlInnodbCluster(AuxiliaryApplication):
 @AppFactory.register_application(["ceph-osd"])
 class CephOSD(AuxiliaryApplication):
     """Application for ceph-osd."""
+
+    def _are_collocated(self, units: list[COUUnit], charm: str) -> bool:
+        """Check if ceph-osd units are collocated with charm.
+
+        :param units: Units to generate upgrade plan
+        :type units: Optional[list[COUUnit]]
+        :param charm: Name of charm
+        :type charm: str
+        :return: Boolean if ceph-osd is colocated with charm.
+        :rtype: bool
+        """
+        return any(
+            # pylint: disable=no-member
+            self.model.applications[app].charm == charm  # type: ignore[attr-defined]
+            for unit in units
+            for app in unit.machine.apps
+        )
+
+    def pre_upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[COUUnit]]
+    ) -> list[PreUpgradeStep]:
+        """Pre Upgrade steps planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate upgrade plan
+        :type units: Optional[list[COUUnit]]
+        :return: List of pre upgrade steps.
+        :rtype: list[PreUpgradeStep]
+        """
+        steps = super().pre_upgrade_steps(target, units)
+        if units is not None and self._are_collocated(units, "nova-compute"):
+            steps.append(
+                PreUpgradeStep(
+                    description="Check if colocated nova-compute units had been upgraded.",
+                    coro=self._verify_nova_compute(target, units),
+                )
+            )
+        return steps
+
+    async def _verify_nova_compute(self, target: OpenStackRelease, units: list[COUUnit]) -> None:
+        """Check if a nova-compute application has upgraded its workload version.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to check if got upgraded
+        :type units: list[COUUnit]
+        :raises ApplicationError: When collocated nova-compute apps workload version isn't reached.
+        """
+        units_not_upgraded = []
+        apps = await self.model.get_applications()  # reload all applications information
+
+        for unit in units:
+            for name in unit.machine.apps:
+                app = apps[name]
+                if app.charm == "nova-compute":
+                    for nova_unit in app.units:
+                        compatible_os_versions = OpenStackCodenameLookup.find_compatible_versions(
+                            app.charm, nova_unit.workload_version
+                        )
+
+                        if target not in compatible_os_versions:
+                            units_not_upgraded.append(unit.name)
+
+        if units_not_upgraded:
+            raise ApplicationError(
+                f"Units '{', '.join(units_not_upgraded)}' did not reach {target}."
+            )
```

Comment:
On theory, I think this shouldn't be necessary because every application is responsible to check if it got upgraded or not. E.g: we don't check on keystone if ceph-mon units got upgraded.

However, there is more complexity in this case because the user can run something like `cou upgrade` and some hypervisors are not empty so it doesn't get upgraded. I see some possible solutions for this case:

1. We could check if the hypervisors_machines selected are enough to finish the nova-compute upgrade and then generate ceph-osd or not depending on this. This might be more complex to implement, but I think it will generate a cleaner plan
2. Have this check as post upgrade step on nova-compute class. Generate this post-upgrade depending on the command. E.g: `cou upgrade hypervisors` should not have it
3. Add on ceph-osd as pre-strep as you are doing.

I prefer option `1`, but if we keep the current approach `3` I don't see the need to check if it's colocated or not. We can have a simple check to see if nova-compute units got upgraded. It's indifferent if they are together in the same machine in the upgrade sequence. Moreover, I think this step should be the first one and the others should show a dependency from it.